nav2_waypoint_follower/include/nav2_waypoint_follower/waypoint_follower.hpp

```diff
@@ -115,16 +126,62 @@ class WaypointFollower : public nav2_util::LifecycleNode
    */
   void goalResponseCallback(const rclcpp_action::ClientGoalHandle<ClientT>::SharedPtr & goal);
 
-  // Our action server
+  /**
+   * @brief send robot through each of GPS
+   *        point , which are converted to map frame first then using a client to
+   *        `FollowWaypoints` action.
+   *
+   * @param waypoints, acquired from action client
+   */
+  void followGPSWaypoints();
+
+/**
+ * @brief given some gps_waypoints, converts them to map frame using robot_localization's service `fromLL`.
+ *        Constructs a vector of stamped poses in map frame and returns them.
+ *
+ * @param gps_waypoints
+ * @param parent_node
+ * @param fromll_client
+ * @return std::vector<geometry_msgs::msg::PoseStamped>
+ */
+  static std::vector<geometry_msgs::msg::PoseStamped> convertGPSWaypointstoPosesinMap(
+    const
+    std::vector<sensor_msgs::msg::NavSatFix> & gps_waypoints,
+    const rclcpp_lifecycle::LifecycleNode::SharedPtr & parent_node,
+    const rclcpp::Client<robot_localization::srv::FromLL>::SharedPtr & fromll_client);
+
+  /**
+   * @brief Action client result callback
+   * @param result Result of action server updated asynchronously
+   */
+  void GPSResultCallback(const rclcpp_action::ClientGoalHandle<ClientTGPS>::WrappedResult & result);
+
+  /**
+   * @brief Action client goal response callback
+   * @param goal Response of action server updated asynchronously
+   */
+  void GPSGoalResponseCallback(const rclcpp_action::ClientGoalHandle<ClientTGPS>::SharedPtr & goal);
+
+  // Our action server for waypoint following
   std::unique_ptr<ActionServer> action_server_;
   ActionClient::SharedPtr nav_to_pose_client_;
-  rclcpp::Node::SharedPtr client_node_;
+  rclcpp::Node::SharedPtr nav_to_pose_client_node_;
   std::shared_future<rclcpp_action::ClientGoalHandle<ClientT>::SharedPtr> future_goal_handle_;
   bool stop_on_failure_;
   ActionStatus current_goal_status_;
   int loop_rate_;
   std::vector<int> failed_ids_;
 
+  // Our action server for GPS waypoint following
+  std::unique_ptr<ActionServerGPS> gps_action_server_;
+  rclcpp::Client<robot_localization::srv::FromLL>::SharedPtr from_ll_to_map_client_;
+  ActionClientGPS::SharedPtr waypoint_follower_action_client_;
+  rclcpp::Node::SharedPtr waypoint_follower_client_node_;
```

Comment:
You shouldn't require another node. The WP follower as is works so this should work with that as well. We can safely assume the WP follower and GPS WP follower will not be called at the same time (because that would be nonsensical) so they can share the same node.